### DISCUSSION TITLE
"Used By" on virtual methods in abstract classes return zero results

### DISCUSSION THREAD

**Discussion by ilyvion** [icsharpcode/ILSpy#3213] [Help]
This is the case even if the method is used directly on the abstract type elsewhere in code. I specifically need to locate a concrete class that *overrides* the method and analyze that before getting results in the "Used By" section. Its section will also get me uses of the virtual/abstract parent method.

Is this expected/intended behavior or a bug? I'm using **ILSpy version 9.0.0.7618-preview1**

Here's an example of this behavior:
![image](https://github.com/icsharpcode/ILSpy/assets/767490/075ef3f9-c784-414c-a0c7-d28487336c99)

And here's some code from that `DoNeeds` method that uses the abstract `Need` type directly, not `Need_Chemical_Any` specifically:
![image](https://github.com/icsharpcode/ILSpy/assets/767490/14febcfd-35d9-4785-9b8b-5ff55fd65240)

**Comment by ilyvion** [comment]
Still a problem in **ILSpy version 9.0.0.7660-preview2**. Also a bit frustrating that nobody is answering these posts.

Assuming the method has been overridden, a simple workaround is to open the "overridden by" section and then analyze that one:
![image](https://github.com/icsharpcode/ILSpy/assets/767490/b8ac04f4-063b-45e1-be86-a7b3392027c5)

This only works if something has overridden the method though; if the method is virtual but nothing overrode it, you're shit out of luck.

**Reply by siegfriedpammer** [reply]
Can you please provide an example binary for me to check? Please be aware that I am "out of office" until Tuesday so I won't be able to take a look immediately - I am answering from my phone. Thanks!

**Reply by daiplusplus** [reply]
@siegfriedpammer A good example is `System.Collections.ObjectModel.Collection<T>` and its subclass `System.Collections.ObjectModel.ObservableCollection<T>` (in .NET Framework 4.8, `Collection<T>` is in `mscorlib.dll` and `ObservableCollection<T>` is in `System.dll` (in the GAC).

* `Collection<T>` has `protected virtual void RemoveItem(Int32)`.
* `ObservableCollection<T>` overrides it with `protected override void RemoveItem(Int32)`
* When using ILSpy `9.1.0.7988+03b7444943e720b3134d296c0c8dd3876f8ea4ce`:
  * The Analyzer tool reports zero **Used By** items for `ObservableCollection<T>.RemoveItem`.
  * The Analyzer tool reports 11 **Used By** items for `Collection<T>.RemoveItem`.

Ideally, for `ObservableCollection<T>.RemoveItem(Int32)`, I'd like to see those 11 call-sites reported under a new Analyzer node: "Base Used By", so as to differentiate between nonvirtual calls to `ObservableCollection<T>.RemoveItem` vs virtual calls to `Collection<T>.RemoveItem` (though IIRC, C# won't actually let you make a nonvirtual call to a virtual method? Please correct me if I'm wrong).

I'd also like to see another Analyzer node for `ObservableCollection<T>.RemoveItem(Int32)`, which would be a nonexpanding node named "Overriddes: System.Collections.ObjectModel.Collection<T>.RemoveItem(Int32)` that navigates to the base method directly when double-clicked.

(...I appreciate that I haven't thought-though how virtual methods with multiple overrides in a 2-or-more inheritance hiearchy should be handled (how hard could it be? 😺))

...and the same also applied to overridden interface members, if possible.

------------

BTW, I noticed the nodes under "Used By" and "Overridden By" don't seem to be sorted (or are they sorted by assembly _then_ by full-name?), can we get them sorted _only_ by full-name - or add an option to have intermediate grouping nodes for each containing-assembly?

<img width="1690" height="1287" alt="image" src="https://github.com/user-attachments/assets/e8ca9029-2af7-4c92-af06-438f15ca1741" />
